src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Sources/ManualResetValueTaskSourceCore.cs

```diff
@@ -265,6 +286,36 @@ private void InvokeContinuation()
                     break;
             }
         }
+
+        /// <summary>
+        /// Invokes the continuation inline and captures any exception thrown.
+        /// This assumes that if <see cref="_continuation"/> is not null we're already
+        /// running within that <see cref="ExecutionContext"/>.
+        /// </summary>
+        private ExceptionDispatchInfo? InvokeContinuationInline(Thread currentThread)
+        {
+            Debug.Assert(_continuation != null);
+            Debug.Assert(_capturedContext == null);
+            Debug.Assert(!RunContinuationsAsynchronously);
+
+            ExceptionDispatchInfo? edi = null;
+            SynchronizationContext? syncContext = currentThread._synchronizationContext;
```

Comment:
MRVTSC definitely shouldn't be mucking with state on Thread.

Reply:
That's the weirdness of `SynchronizationContext` not being part of `ExecutionContext`; but also being saved and restored by `ExecutionContext.Run`

Reply:
But such a change is propagating that weirdness into areas of code that shouldn't have access to it.  It's just making it worse, and it's doing something that any other use outside of corelib couldn't.  If we're going to expose a new API here, I'd like MRVTSC to be able to use it, and without requiring access to the bowels of the runtime.

Reply:
Don't particularly care about the `SynchronizationContext` for the use case in ASP.NET; only clearing all AsyncLocals back to their default state; but obviously this is more general purpose api (and is also using  `SynchronizationContext`/`TaskScheduler` paths), so is important here.

Reply:
Sorry, I didn't completely follow that.  But just in case, I'll try to be very clear in my request:

As a prereq of exposing EC.Restore(EC), MRVTSC should use it, and without requiring any additional hacks or internals access to workaround deficiencies in EC.Restore(EC).

Does it incur any measurable perf penalty for common use cases when doing so?  I don't think it should; ExecutionContext should _never_ be captured in the 99.9% use case.

Reply:
>  I'd like MRVTSC to be able to use it, and without requiring access to the bowels of the runtime.

Could return an opaque struct (that contained `ExecutionContext`+`SynchronizationContext`) then restore that? (opaque so doesn't need to do the extra hoops of `ExecutionContext.Capture()` of changing the returned value from the actual value on thread)

Reply:
> Could return an opaque struct (that contained ExecutionContext+SynchronizationContext) then restore that? (opaque so doesn't need to do the extra hoops of ExecutionContext.Capture() of changing the returned value from the actual value on thread)

What problem are you trying to solve? :-)  Have you measured the perf and found it to be lacking and that's why this is an issue?  Would that opaque struct help with the ASP.NET use case in some way?

Reply:
> What problem are you trying to solve?

To undo any changes running inline has on the sync context.

> Would that opaque struct help with the ASP.NET use case in some way?

It would help with `MRVTSC` because it wouldn't need to mess with sync context directly.

It wouldn't help with ASP.NET because it doesn't care about the return value (it's one way, just trying to restore single previously captured context) or the sync context.

Whereas `MRVTSC` does operate as a stack; restore captured context, do work, restore current context.

ASP.NET is more: do work, restore context (because I don't trust the work to guard asynclocal changes inside `await`s); i.e. restore clean context at end of request loop ready for next request (and there are two examples of it going wrong this way dotnet/aspnetcore#15384 and dotnet/aspnetcore#13991)

The current ASP.NET solution is to move the 3 user calls out behind a new `async` method and let `AsyncMethodBuilderCore.Start<TStateMachine>` handle returning the ExecutionContext back to default.

However, that both introduces a new async statemachine and an extra allocation if any of the methods suspend and its a bit heavyweight just to reset the EC. 

So next thought would be to guard them with `EC.Run`; but then what Task to return and await, since there are 3 of them; if they all complete sync, then can in some way return a CompletedTask; but that's complicated since its callback based (e.g. via the `ref` param); but if one goes async other than the very last one, suddenly it becomes complicated; etc.

If there is a `EC.Restore` on the other hand, can just inline all the awaits into the single method; and don't need an extra startmachine to handle resetting the EC at end of request and don't allocate anything extra if they suspend (since the loop statemachine is already allocated as it likely suspends between requests waiting for the next one)

Reply:
> It would help with MRVTSC because it wouldn't need to mess with sync context directly.

But it doesn't need to mess with sync ctx directly either if uses EC.Restore(EC), no?  I'll come back to my earlier statement:

"Does it incur any measurable perf penalty for common use cases when doing so? I don't think it should; ExecutionContext should never be captured in the 99.9% use case."

When MRVTSC is used with awaits, it won't capture ExecutionContext; it'll only capture EC if you manually use ValueTask.GetAwaiter().OnCompleted (rather than UnsafeOnCompleted).  So all this code around manually managing synchronization context and avoiding multiple Thread lookups is for a code path that will practically never be used, no?

And even if it did matter, exactly how much are we paying for it?  Seems like it's one additional ThreadStatic read on a code path that's going to be reading and writing multiple fields, invoking delegates, using interlocked operations, etc.  If you just do the "simple" thing with the ExecutionContext.Restore that you'd want exposed publicly, and you exercise the code path with ValueTask.GetAwaiter().OnCompleted... do you actually measure a difference between that and all this fancy stuff?

Reply:
> When MRVTSC is used with awaits, it won't capture ExecutionContext;

Ah ok; then it probably doesn't make sense to optimize it :)

> So all this code around manually managing synchronization context 

This however is still a problem; unless don't care about an inline execution changing SyncContext. The issue is SyncContext isn't captured by ExecutionContext; but its changes are reverted by EC.Run; so to mimic that behaviour it needs an extra step.

An alternative would be to wrap the single path that does it in the `EC.Run<T>(ref T)`; but then we still have that rather large generic method just to avoid capturing and restoring the SyncContext.

Reply:
> This however is still a problem

Doesn't that speak to a general problem with the proposed API?  Anyone using it is going to face a similar problem, and doesn't have the ability to access corelib internals to work around it.

Reply:
Can just use the public apis:
```csharp
SynchronizationContext syncCtx = SynchronizationContext.Current;

// do stuff

SynchronizationContext.SetSynchronizationContext(syncCtx);
```

Reply:
Ok. Then MRVTSC should, too. I don't want it reaching into thread internals.

Reply:
Done